cpp/cmake_modules/BuildUtils.cmake

```diff
@@ -165,6 +165,8 @@ function(ADD_ARROW_LIB LIB_NAME)
       LIBRARY_OUTPUT_DIRECTORY "${BUILD_OUTPUT_ROOT_DIRECTORY}"
       OUTPUT_NAME ${LIB_NAME_STATIC})
 
+  target_compile_definitions(${LIB_NAME}_static PUBLIC ARROW_STATIC)
```

Comment:
I'm not sure what this does. Does this only impact the link step for the static library? Are the .cc units compiled a single time or multiple times (once for static, once for shared)?

Reply:
The .cc files are compiled separately for the shared and static library targets (${LIB_NAME}_shared and ${LIB_NAME}_static, respectively). The added line defines the preprocessor macro ARROW_STATIC when compiling sources for the static library.

Reply:
Hi @JohnPJenkins , I also have doubts that we have one time compilation currently of .cc files into obj files and later these obj files are linked both into shared and static libs.

Reply:
Here https://github.com/apache/arrow/blob/master/cpp/cmake_modules/BuildUtils.cmake#L101 we create [Object Library](https://cmake.org/cmake/help/v3.8/manual/cmake-buildsystem.7.html#object-libraries) and it seems use it later both for static and shared libs.

Reply:
@JohnPJenkins also on Windows if you will specify `NMake Makefiles` generator you should be able pretty easy count in console output compilation log how many times the same source files were built.

Reply:
You're right, I missed that one - this patch won't work then when building both libraries. Perhaps I should update the PR to build separate object libraries for the shared and static variants?

Reply:
@JohnPJenkins it seems logic should be changed only for Windows (to not increase compilation time on Unix). And on Windows maybe it makes sense to build directly from sources in cmake script, since it's not possible to reuse object files.

Reply:
The objlib thing is an optimization for Unix/macOS so this part could be skipped on Windows

Reply:
That makes sense - looking more closely at the cmake file, the unix builds are unconditionally using PIC, so no issues there.